src/gui/PianoView.cpp

```diff
@@ -402,33 +417,59 @@ void PianoView::pianoScrolled(int new_pos)
 }
 
 
-
-
-/*! \brief Handle a context menu selection on the piano display view
+/*! \brief Handle a context menu selection on the piano view
  *
  *  \param me the ContextMenuEvent to handle.
  *  \todo Is this right, or does this create the context menu?
  */
 void PianoView::contextMenuEvent(QContextMenuEvent *me)
 {
-	if (me->pos().y() > PIANO_BASE || m_piano == nullptr ||
+	if (m_piano == nullptr ||
 		m_piano->instrumentTrack()->keyRangeImport())
 	{
 		QWidget::contextMenuEvent(me);
 		return;
 	}
 
-	// check which control element is closest to the mouse and open the appropriate menu
 	QString title;
-	IntModel *noteModel = getNearestMarker(getKeyFromMouse(me->pos()), &title);
+	int key_num = getKeyFromMouse(me->pos());
 
-	CaptionMenu contextMenu(title);
-	AutomatableModelView amv(noteModel, &contextMenu);
-	amv.addDefaultActions(&contextMenu);
-	contextMenu.exec(QCursor::pos());
-}
+	if (me->pos().y() > PIANO_BASE)
+	{
+		// context menu for the key area
+		m_lastContextMenuKey = key_num;
```

Comment:
Rather than passing this to the various menu handlers via a member variable, I think it would be better to pass it as a normal parameter. You can do this by connecting the menu item to a lambda and capturing `key_num`:
```cpp
contextMenu.addAction(tr("Foo"), this, [this, key_num]() { setFoo(key_num); });
```

Reply:
So it would be possible to write the following without the SLOT macro or would it still be needed somehow?

```c++
QAction *actionFirst = contextMenu.addAction(tr("Set &first key"), this, [this, key_num]() { setFirstKey(key_num); });
```

Reply:
That should work, no `SLOT` macro required. It's even possible to move `setFirstKey` and the others out of `private slots:` and into a normal `private:` section, since they're only called as normal functions now.

(The `SLOT` macro just converts its argument into a string, which `addAction` passes to `connect` internally, where it is used to look up the actual slot at runtime. This is a legacy style from Qt4; Qt5 supports passing a slot directly as a member function pointer, or passing an object implementing `operator()` such as a lambda.)